### DISCUSSION TITLE
Negative inH2O values? and different full scale reading between Depression and Diff/Pitot.

### DISCUSSION THREAD

**Discussion by KirikauKiwi** [DIY-Flow-Bench/DIY-Flow-Bench#191] [Code Development]
Thinking aloud here as just having got my pressure sensors up and running today and adjusting the millivolt trim for them, I noticed that if I applied a to large of a negative offset then the inH2O would essential be locked to zero due to it not reading negatively.
Would it be useful to allow for negative inH2O to see this, could be useful too, incase the sensors drift or some other oddity happens in the system.

I also noticed due to my sensors being -100 to 200kPA that I can get a full scale reading on Depression of 865 inH2O ≈ 215kPa, when I do the same test on Diff Pressure it only reads 219 inH2O ≈ 54kPA this is also the same for Pitot. Is there any math conversion done on Diff Pressure and Pitot other than the ADC, or is it just a display limit?

**Comment by DeeEmm** [comment]
Yes the value should ideally swing both ways to allow for reverse flow. I will take a look at this

Not sure about the second question, will need to take a look. :)

**Comment by DeeEmm** [comment]
As I mentioned int he othrr discussion, it's the following code...

```
  // Flow Rate
  if ((sensorVal.FlowCFM > config.min_flow_rate) && (sensorVal.PRefH2O > config.min_bench_pressure))  {
    dataJson["FLOW"] = sensorVal.FlowCFM;
    dataJson["MFLOW"] = sensorVal.FlowKGH;
  }  else  {
    dataJson["FLOW"] = 0.0;
    dataJson["MFLOW"] = 0.0;
  }
```

We are only testing to see if the value is greater than the threshold. to display negative flow values we also need to check that it is below the negative threshold. So we should be able to multiply the value by '-1' to change the sign for the comparison.

**Comment by DeeEmm** [comment]
This should hopefully work...

```

  double flowComp = sensorVal.FlowCFM * -1;
  double pRefComp = sensorVal.PRefH2O * -1;

  // Flow Rate
  if ((flowComp > config.min_flow_rate) && (pRefComp > config.min_bench_pressure))  {
    dataJson["FLOW"] = sensorVal.FlowCFM;
    dataJson["MFLOW"] = sensorVal.FlowKGH;
  }  else  {
    dataJson["FLOW"] = 0.0;
    dataJson["MFLOW"] = 0.0;
  }
```

**Comment by DeeEmm** [comment]
<img width="783" alt="Screenshot 2024-11-03 at 5 52 07 pm" src="https://github.com/user-attachments/assets/d0413900-018a-4b27-89c7-70ecab82aa6c">

OK. multiplying by -1 did not work for some reason. unsure why. however using fabs() worked fine

Just tidying up my working branch and I will create a PR for the changes.

**Reply by KirikauKiwi** [reply]
Does the current Build 24110403 include the negative numbers?, I am running this as the latest build from your master. I can see the changes to faps( ) in the hardware and webserver's cpp, however when I apply a vacuum the display is still stopping at zero.

Did you find anything relating to the PDiff and Pitot not reading the full scale of the sensors, I have looked through DIYFB codebase, sensors, hardware and ADS1115_lite etc and cannot see anything obvious. I see that the ads is configured for single-ended on all four channels so I would expect to get the full scale reading as the three pressure sensor are all the same in the code base other than prefix in the naming of them.
I had missed adding the code for the PRef, PDiff and Pitot in the void Sensors::initialise () { part of the sensors.cpp, having added and updated that I still don't get the full scale reading, which I wouldn't expect it to have changed anything as I had missed it for all three sensors, but good that I pick up the missing bits.

It behaves as if the sensor voltage is maxing out on all three pressure channels. I have downloaded CoolTerm so will connect that up and see if I can see anything, are there API codes for PDiff and Pitot?

Are you able to get full scale readings for PDiff and Pitot with the DIYFB standard configuration i.e Wemos and MXVP7007 or do you see the same as I am. This would be helpful as it would either show that it is something I have missed in adding the M5Stack sensors to the code base or if it is happening for all of us.

**Comment by KirikauKiwi** [comment]
Ok, with a bit more testing and further understanding I can see now that I get the negative numbers in relation to what would be reverse flow.
I still don't see a negative number relating to the readings inH20 for Pref, PDiff and Pitot. e.g. if I apply a vacuum to the pressure side or a pressure to the vacuum side of the sensor should it not provide a -inH2O on the GUI due to the voltage offset of 2.5 v for the MPXV7007 sensor, with 0V to 2.5V = -7kPa to 0 kPa and 2.5V to 5V = 0kPa to +7kPa, currently if a vacuum is present on the pressure port the GUI reading stops at 0.00 inH2O.

This is due to the fact the sensors I am using are single sided and how DIYFB uses the dual sided sensors currently.

As the depression is measured as a vacuum and the line is connected to the lower port on the sensor, so in this instance for the default hardware you would not be able to measure a positive pressure which would display on the GUI as -inH20, a negative depression would be a positive pressure (double negative = positive).

If this was implemented in to DIYFB I think it would allow for exhaust port testing with out having to switch the vacuum lines across to the other side of the sensor when reversing the flow of the motors and there fore making DIYFB easier to use when switching between inlet and exhaust testing, you would only need to flip the orientation of the MAF sensor.

TLDR as above was me brain dumping, essentially we have a single sided manometer in DIYFB currently

Update: 
Testing with a bench top PSU the depression is being measured from 2.5V to 5V which makes sense as the way the a flow bench is connected to the lower port on the sensor as this would be the same as applying a positive pressure to the top port. 
When inputing 0V to 2.5V from the PSU to the ADS1115 the GUI displays 0.0 inH2O.
In my mind the GUI depression should be reading in the -7kPA region of the sensor which is 0V to 2.5V range. If the Bench was connected to the top port of the sensor this would allow for the full range of the sensor to be utilised and allow negative and positive pressures readings to be made

**Reply by DeeEmm** [reply]
Yes the ability to measure  +/- flow without having to re-pipe the sensors was the intent and should work. The value decoding should be taken care of by the transfer function... zero is actually about 2.2ish volts, below this the pressure is negative.

<img width="712" alt="Screenshot 2024-11-06 at 11 11 31 pm" src="https://github.com/user-attachments/assets/c526db10-9cb8-4625-9bd5-7f4d5968a335">

I am unfortunately without an internet connection at the moment as a lightening strike yesterday fried both our HFC box and modem. I'm waiting for a tech to come out.

I think that the issue is the same as previously - the value is zeroed if the flow is zero. (schoolboy error lol)

```
	// Lets make sure we have a valid value to return
	if (sensorVal > 0) { 
		return sensorVal;
	} else { 
		return 0.0001; // return small non zero value to prevent divide by zero errors (will be truncated to zero in display)
	}
	
```

The solution is the same as before. Convert to absolute using fabs before we compare.  If it is below the threshold, then return zero. We could also possibly just remove this test altogether.

I note my comment about a Divide By zero error, but I'm not sure if this is still applicable. Originally the code was much different.

```
	// Lets make sure we have a valid value to return
	double pRefComp = fabs(sensorVal);
	if (pRefComp > settings.min_bench_pressure) { 
		return sensorVal;
	} else { 
		return 0.0001; // return small non zero value to prevent divide by zero errors (will be truncated to zero in display)
	}
```

I am unable to test this at the moment but I will roll out a fix when my connection is back up and running.

**Reply by DeeEmm** [reply]
(the same issue exists for pRef, pDiff and pitot)

**Reply by KirikauKiwi** [reply]
Sorry to hear about the lightning strike, enjoy the down time.

With the MPXV7007DP sensor are you connected to the port closest to the PCB with the upper port open to atmosphere.?

**Reply by DeeEmm** [reply]
Sorry I don't recall. At the moment I only have a test board set up with pots instead of sensors for bench testing. I need to get my flow bench set up again and do some testing.

The lightening strike was crazy. Was in my office, heard a small electrical pop, like a fuse blowing, and a flash out of the corner of my eye where the modem is, and then a split second later a massive lightening crash that seemed to be right outside my window. I'm guessing it must have hit the NBN line somewhere. Super frustrating tho.

**Reply by KirikauKiwi** [reply]
I had a go with the code change you suggested above.

```
// Lets make sure we have a valid value to return
	double pRefComp = fabs(sensorVal);
	if (pRefComp > settings.min_bench_pressure) { 
		return sensorVal;
	} else { 
		return 0.0001; // return small non zero value to prevent divide by zero errors (will be truncated to zero in display)
	}
```

This is what happened while I compiled it.

```
ESP32/DIY-Flow-Bench/sensors.cpp: In member function 'double Sensors::getPRefValue()':
ESP32/DIY-Flow-Bench/sensors.cpp:674:18: error: 'settings' was not declared in this scope
   if (pRefComp > settings.min_bench_pressure) {
                  ^~~~~~~~
ESP32/DIY-Flow-Bench/sensors.cpp:674:18: note: suggested alternative: 'sbttons'
   if (pRefComp > settings.min_bench_pressure) {
                  ^~~~~~~~
                  sbttons
ESP32/DIY-Flow-Bench/sensors.cpp:673:9: warning: control reaches end of non-void function [-Wreturn-type]
  double pRefComp = fabs(sensorVal);
```

I tried to work through and understand the compile errors, unfortunately I couldn't fix it, looking forward to seeing the changes you make. 😄

**Comment by KirikauKiwi** [comment]
I think there are two things in play here, one physical and one code based as you mentioned above.

The physical one is the flow bench would need to be connected to P1 in the picture below, when a pressure is applied it causes the die to deflect down giving the 2.2V to 5V signal, if a vacuum is applied to P1 the die will be deflect up towards the top giving a decreasing signal from 2.2V to 0V.
If the flow bench is connected to P2 and the vacuum applied this causes the it to deflect down which means it is reading a positive pressure on the 2.2V to 5V range.
I think currently the pressure sensor is only being used for half of its available range due to this, because if the flow bench was connected to P1 it would read 0.0 on the DIYFB GUI

<img width="949" alt="Screenshot 2024-11-06 at 13 50 18" src="https://github.com/user-attachments/assets/cc5def63-a171-4fdb-bd46-cd7d7a7bc516">

**Reply by KirikauKiwi** [reply]
Looking at @stefan63 pictures in #174 the vacuum hose is connected to the lower port P2.

**Comment by KirikauKiwi** [comment]
I have found where the difference in full-scale readings are coming from on the DIYFB GUI. I am seeing full scale readings but they are being displayed in different units to what they are labeled on the GUI.

<img width="709" alt="Screenshot 2024-11-06 at 18 07 37" src="https://github.com/user-attachments/assets/68e7d60a-b426-41ec-863c-3a8253601f7d">

Depression is being displayed as inH20 as it is labeled.
Diff pressure is being displayed in kPa but is labeled as inH2O
Pitot is being displayed in kPa but is labeled as in H2O.

Found this in the webserver.cpp
```
// Pitot
  dataJson["PITOT"] = sensorVal.PitotKPA;

  // Differential pressure
  dataJson["PDIFF"] = sensorVal.PDiffKPA;
```
there appears to be no reference for sensorVal.PitotH2O or PDiffH2O in the structs.h

**Reply by DeeEmm** [reply]
Good pickup!

Fixed. I've updated changes to master.

kPa is the default unit used in most of the calcs. Somehow it's been overlooked for the pDiff and Pitot values sent to the UI.

**Comment by KirikauKiwi** [comment]
Updated to Build 24110403

It appears that DIYFB is somehow using 2.2V - 5V  to measure the depression, should it not be using the 0V - 2.2V to measure depression.

Here is what I am getting on the GUI currently my sensor is 0.1V - 1.1V for -100kPA to 0kPA and 1.1V - 3.3V for 0kPA to 200Kpa.

**PRef approx 3.3V my 200 kPa ≈ 885 inH2O**
<img width="743" alt="Screenshot 2024-11-07 at 12 58 46" src="https://github.com/user-attachments/assets/1eb4cdd9-4852-4d1f-81da-545a19be0f3c">


**PRef approx 1.1V my zero kPa ≈ 0 inH2O**
<img width="743" alt="Screenshot 2024-11-07 at 12 46 41" src="https://github.com/user-attachments/assets/b1624212-9218-4557-a6fe-c9dae10332ed">


**PRef approx 0.3V my -100 kPa ≈ 401 inH2O**
<img width="743" alt="Screenshot 2024-11-07 at 12 47 31" src="https://github.com/user-attachments/assets/5327b8b1-c258-4dfa-8531-96296d679dc0">

**Reply by KirikauKiwi** [reply]
The transfer functions appear to be working correctly as the reading I get from Cool Term is following it correctly.
The display in the DIYFB GUI is giving me a numeric value that is a pressure reading according to the transfer function. When I cross from 1.1V or below the DIYFB GUI is reading 0.00 when I should be getting a valid Depression value.

Is it possible to make the "Depression" heading automatically switch between "Pressure" and "Depression" when voltage is above or below the Zero kPa crossing point respectively?

**Comment by DeeEmm** [comment]
Also rolled out the fix for the negative pressure values. Let me know how it works for you. ATM i've got the dev/debug board set up to look at the stuff on the WIP branch so unable to test

**Reply by KirikauKiwi** [reply]
Looks all good now. 👍 for the negative pressures and the kPa to inH2O GUI changes

<img width="743" alt="Screenshot 2024-11-07 at 13 39 09" src="https://github.com/user-attachments/assets/92679069-b3b8-432b-aff9-a678d09b4058">
